File: bull-common/src/test/java/com/hotels/transformer/base/DefaultsTest.java
/**
 * Copyright (C) 2019-2021 Expedia, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hotels.transformer.base;

import static java.lang.Boolean.FALSE;

import static org.assertj.core.api.Assertions.assertThat;

import org.mockito.InjectMocks;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import com.hotels.beans.sample.FromFoo;

/**
 * Unit test for {@link Defaults}.
 */
public class DefaultsTest {
    /**
     * The class to be tested.
     */
    @InjectMocks
    private Defaults underTest;

    /**
     * Tests that the method: {@code defaultValue} returns the expected result for the given type.
     * @param type the class type.
     * @param expectedResult the expected result.
     */
    @Test(dataProvider = "dataProvider")
    public void testDefaultValueShouldReturnTheExpectedResult(final Class<?> type, final Object expectedResult) {
        // GIVEN

        // WHEN
        final Object actual = underTest.defaultValue(type);

        // THEN
        assertThat(actual).isEqualTo(expectedResult);
    }

    /**
     * Data provider for the required test cases.
     * @return the test cases.
     */
    @DataProvider
    private Object[][] dataProvider() {
      return new Object[][] {
                {boolean.class, FALSE},
                {Boolean.class, FALSE},
                {Character.class, '\u0000'},
                {char.class, '\u0000'},
                {Byte.class, 0},
                {byte.class, 0},
                {Integer.class, 0},
                {int.class, 0},
                {Short.class, 0},
                {short.class, 0},
                {Long.class, 0L},
                {long.class, 0L},
                {Float.class, 0.0F},
                {float.class, 0.0F},
                {Double.class, 0.0D},
                {double.class, 0.0D},
                {FromFoo.class, null}
      };
    }
}

File: bull-common/src/test/kotlin/com/expediagroup/transformer/base/DefaultsTest.kt
/**
 * Copyright (C) 2019-2022 Expedia, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.expediagroup.transformer.base

import com.expediagroup.beans.sample.FromFoo
import org.assertj.core.api.Assertions
import org.mockito.InjectMocks
import org.testng.annotations.DataProvider
import org.testng.annotations.Test

/**
 * Unit test for [Defaults].
 */
class DefaultsTest {
    /**
     * The class to be tested.
     */
    @InjectMocks
    private val underTest: Defaults? = null

    /**
     * Tests that the method: `defaultValue` returns the expected result for the given type.
     * @param type the class type.
     * @param expectedResult the expected result.
     */
    @Test(dataProvider = "dataProvider")
    fun testDefaultValueShouldReturnTheExpectedResult(type: Class<*>?, expectedResult: Any?) {
        // GIVEN

        // WHEN
        val actual = Defaults.defaultValue(type)

        // THEN
        Assertions.assertThat(actual).isEqualTo(expectedResult)
    }

    /**
     * Data provider for the required test cases.
     * @return the test cases.
     */
    @DataProvider
    private fun dataProvider(): Array<Array<Any?>> {
        return arrayOf(
            arrayOf(Boolean::class.javaPrimitiveType, Boolean.FALSE),
            arrayOf(
                Boolean::class.java,
                Boolean.FALSE
            ),
            arrayOf(Char::class.java, '\u0000'),
            arrayOf(
                Char::class.javaPrimitiveType,
                '\u0000'
            ),
            arrayOf(Byte::class.java, 0),
            arrayOf(
                Byte::class.javaPrimitiveType,
                0
            ),
            arrayOf(Int::class.java, 0), arrayOf(Int::class.javaPrimitiveType, 0),
            arrayOf(
                Short::class.java,
                0
            ),
            arrayOf(Short::class.javaPrimitiveType, 0), arrayOf(Long::class.java, 0L),
            arrayOf(
                Long::class.javaPrimitiveType,
                0L
            ),
            arrayOf(Float::class.java, 0.0f),
            arrayOf(
                Float::class.javaPrimitiveType,
                0.0f
            ),
            arrayOf(Double::class.java, 0.0),
            arrayOf(
                Double::class.javaPrimitiveType,
                0.0
            ),
            arrayOf(FromFoo::class.java, null)
        )
    }
}
